test/functional/lua/fs_spec.lua

```diff
@@ -308,26 +309,119 @@ describe('vim.fs', function()
+          vim.fs.normalize([[\\.\Volume{12345678-1234-1234-1234-1234567890AB}\foo\bar]])
+        )
+      end)
+      it('errors on invalid UNC and DOS device paths', function()
+        eq(
+          '.../fs.lua:0: Invalid Windows UNC path',
+          pcall_err(vim.fs.normalize, [[\\server\share]])
+        )
+        eq('.../fs.lua:0: Invalid Windows UNC path', pcall_err(vim.fs.normalize, [[\\server\]]))
+        eq(
+          '.../fs.lua:0: Invalid Windows UNC path',
+          pcall_err(vim.fs.normalize, [[\\.\UNC\server\share]])
+        )
+        eq(
+          '.../fs.lua:0: Invalid Windows UNC path',
+          pcall_err(vim.fs.normalize, [[\\?\UNC\server\]])
+        )
+        eq('.../fs.lua:0: Invalid Windows UNC path', pcall_err(vim.fs.normalize, [[\\.]]))
+        eq('.../fs.lua:0: Invalid Windows device path', pcall_err(vim.fs.normalize, [[\\.\]]))
+        eq('.../fs.lua:0: Invalid Windows device path', pcall_err(vim.fs.normalize, [[\\.\foo]]))
+        eq(
+          '.../fs.lua:0: Invalid Windows device path',
+          pcall_err(vim.fs.normalize, [[\\.\BootPartition]])
+        )
+      end)
+      it('converts backward slashes', function()
+        eq('C:/Users/jdoe', vim.fs.normalize([[C:\Users\jdoe]]))
+      end)
     end
+
+    describe('. and .. component resolving', function()
+      it('works', function()
+        if is_os('win') then
+          eq('C:/Users', vim.fs.normalize([[C:\Users\jdoe\Downloads\.\..\..\]]))
+          eq('C:/Users/jdoe', vim.fs.normalize([[C:\Users\jdoe\Downloads\.\..\.\.\]]))
+          eq('C:/', vim.fs.normalize('C:/Users/jdoe/Downloads/./../../../'))
+          eq('C:foo', vim.fs.normalize([[C:foo\bar\.\..\.]]))
+        else
+          eq('/home', vim.fs.normalize('/home/jdoe/Downloads/./../..'))
+          eq('/home/jdoe', vim.fs.normalize('/home/jdoe/Downloads/./../././'))
+          eq('/', vim.fs.normalize('/home/jdoe/Downloads/./../../../'))
+        end
+
+        eq('foo/bar/baz', vim.fs.normalize('foo/bar/foobar/../baz/./'))
+        eq('foo/bar', vim.fs.normalize('foo/bar/foobar/../baz/./../../bar/./.'))
+      end)
+
+      it('works when relative path reaches current directory', function()
+        if is_os('win') then
+          eq('C:', vim.fs.normalize('C:foo/bar/../../.'))
+        end
+
+        eq('.', vim.fs.normalize('.'))
+        eq('.', vim.fs.normalize('././././'))
+        eq('.', vim.fs.normalize('foo/bar/../../.'))
+      end)
+
+      it('works when relative path goes outside current directory', function()
+        eq('../../foo/bar', vim.fs.normalize('../../foo/bar'))
+        eq('../foo', vim.fs.normalize('foo/bar/../../../foo'))
+
+        if is_os('win') then
+          eq('C:../foo', vim.fs.normalize('C:../foo'))
+          eq('C:../../foo/bar', vim.fs.normalize('C:foo/../../../foo/bar'))
+        end
+      end)
+
+      it('.. in root directory resolves to itself', function()
+        if is_os('win') then
```

Comment:
Any reason we shouldn't have a `unc:boolean` or `win:boolean` option? That allows (1) tests can run in all environments, and (2) more flexibility for plugins that might need to work with windows/unc paths in a non-windows context.

Reply:
We could add an option like `path_style`, in case someone wants to process a path style that's different from their OS. I do think that should be a separate PR though.

Reply:
Isn't what you're talking about less about unc and more about operating system?

Reply:
> Isn't what you're talking about less about unc and more about operating system?

UNC paths are only processed in Windows

Reply:
I know. I'm saying I think a naming the option `unc` is wack since what Justin is suggesting (normalizing paths for another OS) can be relevant even without UNC, e.g. `C:/Users/../jdoe/.`.

Reply:
> I know. I'm saying I think a naming the option `unc` is wack since what Justin is suggesting (normalizing paths for another OS) can be relevant even without UNC, e.g. `C:/Users/../jdoe/.`.

Yes, that's why I never suggested naming it UNC. I suggested a name like `path_style`, which would be like an enum that can be either `posix` or `win`

Reply:
Will make this in a separate PR